Mcol 4313 macros

drrtuy opened the pull request.

drrtuy committed:
MCOL-4313 Introduced preprocessor macro to efficiently assign 128 bit value
based on a compiler version

drrtuy committed:
MCOL-4313 Refactored assignments in the code so now we use
a templated asignmment function with SIMD-aware macro behind

Another toString() method for wide-decimal that is EMPTY/NULL aware

Unified decimal processing in WF functions

Fixed a potential issue in EqualCompData::operator() for
wide-decimal processing

Removed uint128_t from the code and fixed some signedness warnings

tntnatbry commented on datatypes/mcs_decimal.cpp:
```diff
@@ -209,6 +217,23 @@ namespace datatypes
+            return std::string("EMPTY");
+        }
+        int128_t& constLessValue = const_cast<int128_t&>(value);
+        dataconvert::DataConvert::decimalToString(&constLessValue,
```
I noticed the first parameter to `dataconvert::DataConvert::decimalToString` is declared as int128_t*. I think it should be const int128_t*. Also `decimalToString` is changing the source value from negative to positive. I don't think that's intended.

drrtuy replied:
It should be but outside the scope of the change. And yes, decimalToString has this flaw.

tntnatbry commented on utils/common/simd_asm.h:
```diff
@@ -0,0 +1,77 @@
+                                src, \
+                                src_restrictions, \
+                                clobb) \
+        ::memcpy((dst), (src), sizeof(__int128));
```
This should be `&(src)`.

drrtuy replied:
JFYI The macro will be replaced with the ASM variant from #else preprocessor branch.

tntnatbry commented on utils/rowgroup/rowgroup.h:
```diff
@@ -1165,12 +1152,7 @@ inline void Row::setLongDoubleField(const long double& val, uint32_t colIndex)
-inline void Row::setUint128Field(const uint128_t& val, uint32_t colIndex)
-{
-    *((uint128_t*)&data[offsets[colIndex]]) = val;
+    setBinaryField<int128_t>(&(const_cast<int128_t&>(val)), colIndex);
```
I think the first parameter of `setBinaryField` should be a `const T*`.

drrtuy replied:
Agree.

tntnatbry commented on utils/windowfunction/frameboundrange.cpp:
```diff
@@ -299,21 +299,26 @@ void FrameBoundExpressionRange<T>::validate()
+                                < datatypes::MAXDECIMALWIDTH))
                 {
-                    int64_t tmp = this->fRow.getIntField(this->fIndex[1]);
+                    uint64_t tmp = this->fRow.getIntField(this->fIndex[1]);
```
This is not correct for narrow decimals; the rhs returns an int64_t for an UDECIMAL which is not correct. I think we should keep the original two separate cases for DECIMAL and UDECIMAL.

drrtuy replied:
I didn't get the point to be honest.

drrtuy committed:
MCOL-4313 Replaced template asignment functions with TSInt128 class

Replaced assignment at the points that gets SEGV working with int128_t

Removed uint128 from joblist/lbidlist.*
